File: examples/code_lifecycle/native_code/code_lifecycle.f90
module mod_code_lifecycle

contains

    !
    !    INITIALISATION SUBROUTINE
    !
    subroutine init_code (xml_parameters, status_code, status_message)
        use ids_schemas, only: ids_parameters_input
        implicit none
        type(ids_parameters_input) :: xml_parameters
        integer, intent(out) :: status_code
        character(len=:), pointer, intent(out) :: status_message


        ! Setting status to SUCCESS
        status_code = 0
        allocate(character(50):: status_message)
        status_message = 'OK'

        write(*,*) '======================================='
        write(*,*) 'Code lifecycle: INITIALISATION called'
        write(*,*) '======================================='

    end subroutine init_code


    !
    !    FINALISATION SUBROUTINE
    !
    subroutine clean_up(status_code, status_message)
        implicit none
        integer, intent(out) :: status_code
        character(len=:), pointer, intent(out) :: status_message

        ! Setting status to SUCCESS
        status_code = 0
        allocate(character(50):: status_message)
        status_message = 'OK'

        write(*,*) '======================================='
        write(*,*) 'Code lifecycle: FINALISATION called'
        write(*,*) '======================================='

    end subroutine clean_up

    !
    !    MAIN SUBROUTINE
    !
    subroutine code_lifecycle(equilibrium_in, equilibrium_out, xml_parameters, status_code, status_message)
        use ids_schemas, only: ids_equilibrium, ids_parameters_input, ids_is_valid
        use ids_routines, only: ids_copy

        implicit none

        type(ids_equilibrium):: equilibrium_in,equilibrium_out
        type(ids_parameters_input) :: xml_parameters
        integer, intent(out) :: status_code
        character(len=:), pointer, intent(out) :: status_message

        ! INITIALISATION OF ERROR FLAG
        status_code = 0

        ! INITIAL DISPLAY
        write(*,*) ' '
        write(*,*) '======================================='
        write(*,*) 'START OF PHYSICS CODE'

        ! CHECK IF INPUT IDS IS VALID
        if (ids_is_valid(equilibrium_in%ids_properties%homogeneous_time).eqv..True. &
            .and.size(equilibrium_in%time)>0) then

            ! COPY THE INPUT IDS IN THE OUTPUT IDS
            call ids_copy(equilibrium_in, equilibrium_out)

            ! MANDATORY FLAG (UNIFORM TIME HERE)
            equilibrium_out%ids_properties%homogeneous_time = 1
            allocate(equilibrium_out%code%name(1))   ! For a string of 132 characters max.

            equilibrium_out%code%name(1)   = 'EXAMPLE: code_lifecycle'
            allocate(equilibrium_out%code%version(1))   ! For a string of 132 characters max.

            equilibrium_out%code%version(1)   = '1.0'
            allocate(equilibrium_out%code%parameters(1))   ! For a string of 132 characters max.
            equilibrium_out%code%parameters(1) = 'EXAMPLE: code_lifecycle: NO PARAMETERS'
            allocate(equilibrium_out%code%output_flag(1))
            equilibrium_out%code%output_flag(1) = 0   ! Integer output flag, 0 means the run was successful and can be used in the rest of the workflow, <0 means failure

        else
            ! ERROR IF THE CODE DOES NOT COMPLETE TO THE END
            status_code = -1
            allocate(character(50):: status_message)
            status_message = 'Error in physics_ii: input IDS not valid'
        endif

        ! FINAL DISPLAY
        write(*,*) 'END OF PHYSICS CODE'
        write(*,*) '======================================='
        write(*,*) ' '

        end subroutine code_lifecycle
end module mod_code_lifecycle

File: examples/code_lifecycle_cpp/native_code/code_lifecycle.cpp
#if 5 == AL_MAJOR
    #include "ALClasses.h"
#elif AL_MAJOR == 4
    #include "UALClasses.h"
#else
    #warning Could not find AL_MAJOR variable. Assuming AL version = 5.x.x
    #include "ALClasses.h"
#endif

#include "code_lifecycle.h"


// =======================================
//             INITIALISATION
//=======================================

void init_code (IdsNs::codeparam_t codeparam,
                int& status_code, std::string& status_message)
{
    status_code = 0;

   status_message = "INITIALISATION: OK";

    printf("=======================================================\n");
    printf("Code lifecycle CPP: INITIALISATION called\n");
    printf("=======================================================\n");
}

// =======================================
//             FINALISATION
//=======================================
void clean_up( int& status_code, std::string& status_message)
{
    status_code = 0;
    status_message = "FINALISATION: OK";

    printf("=======================================================\n");
    printf("Code lifecycle CPP: FINALISATION called\n");
    printf("=======================================================\n");
}

// =======================================
//             MAIN
//=======================================
void code_lifecycle(const IdsNs::IDS::equilibrium& in_equilibrium, IdsNs::IDS::equilibrium& out_equilibrium,
                     IdsNs::codeparam_t xml_params,
                     int& status_code, std::string& status_message)
{
    int idsSize = -1;
    int idsTimeMode = IDS_TIME_MODE_UNKNOWN;
    
    // INITIALISATION OF ERROR FLAG
    status_code = 0;
    
    // INITIAL DISPLAY
    printf( "\n=======================================\n");
    printf( "START OF PHYSICS CODE\n");
    
    // CHECK IF INPUT IDS IS VALID
    idsSize = in_equilibrium.time.extent(0);
    idsTimeMode = in_equilibrium.ids_properties.homogeneous_time;
    if ( idsTimeMode != IDS_TIME_MODE_HOMOGENEOUS && idsSize > 0)
    {   
        // ERROR IF THE CODE DOES NOT COMPLETE TO THE END
        status_code = -1;
        status_message = "Error in code_lifecycle: input IDS not valid";
        return;
    }
    
    // MANDATORY FLAG (UNIFORM TIME HERE)
    out_equilibrium.ids_properties.homogeneous_time = IDS_TIME_MODE_HOMOGENEOUS;
    out_equilibrium.time.resize(idsSize);

    // Fill in the output IDS (Physical data)
    for(int i=0; i < idsSize; i++)
    {
        // Time : copy from input IDS
        out_equilibrium.time(i) = in_equilibrium.time(i);
    }

    out_equilibrium.ids_properties.homogeneous_time = 1;
    out_equilibrium.code.name   = "coreprofiles2distsource_cpp";
    out_equilibrium.code.version   = "1.0";
    out_equilibrium.code.parameters = "my_code_specific_parameters";
    out_equilibrium.code.output_flag.resize(1);
    out_equilibrium.code.output_flag(0) = 0;

    printf("Size of input IDS  = %d\n", idsSize);

    // INITIALISATION OF STATUS INFO
    status_message = "Status info of code_lifecycle CPP";

    
    // FINAL DISPLAY
    printf( "END OF PHYSICS CODE\n");
    printf( "=======================================\n");
    printf( " \n");
}
